File: web/src/lib/gamification.ts
// Android'dagi analysis/Gamification.kt ning aynan porti.
//
// MUHIM: Kotlin'da Int / Int butun sonli bo'lish — shuning uchun bu yerda ham
// Math.floor. Ikkala platformada bir xil daraja va XP chiqishi kerak.
// Bu ikki fayl doim birga o'zgartiriladi.

import type { StoredAttempt } from "./attempts-store";

export type Badge = {
  id: string;
  emoji: string;
  title: string;
  hint: string;
  unlocked: boolean;
};

export type GameStats = {
  totalXp: number;
  level: number;
  levelTitle: string;
  xpInLevel: number;
  xpPerLevel: number;
  streakDays: number;
  bestStreak: number;
  practicedToday: boolean;
  totalAttempts: number;
  totalWords: number;
  badges: Badge[];
  levelProgress: number;
  unlockedBadges: number;
};

const XP_PER_LEVEL = 400;

const LEVEL_TITLES = [
  "Yangi boshlovchi",
  "Mashqchi",
  "Suhbatdosh",
  "Notiq",
  "Usta notiq",
  "Chempion",
];

/** Mahalliy vaqt mintaqasidagi kun raqami (1970-01-01 dan boshlab). */
function dayIndex(timestamp: number): number {
  // getTimezoneOffset shu lahza uchun (DST hisobga olingan holda) daqiqada qaytaradi.
  const offset = -new Date(timestamp).getTimezoneOffset() * 60_000;
  return Math.floor((timestamp + offset) / 86_400_000);
}

/**
 * Joriy seriya: bugundan (yoki kechadan — bugun hali mashq qilmagan bo'lsa)
 * orqaga qarab uzluksiz kunlar soni.
 */
function currentStreak(days: Set<number>, today: number): number {
  let cursor: number;
  if (days.has(today)) cursor = today;
  else if (days.has(today - 1)) cursor = today - 1;
  else return 0;

  let count = 0;
  while (days.has(cursor)) {
    count++;
    cursor--;
  }
  return count;
}

/** Eng uzun uzluksiz kunlar seriyasi (butun tarix bo'yicha). */
function bestStreak(days: Set<number>): number {
  if (days.size === 0) return 0;
  const sorted = [...days].sort((a, b) => a - b);
  let best = 1;
  let run = 1;
  for (let i = 1; i < sorted.length; i++) {
    run = sorted[i] === sorted[i - 1] + 1 ? run + 1 : 1;
    if (run > best) best = run;
  }
  return best;
}

export function computeGameStats(
  attempts: StoredAttempt[],
  moduleCount: number,
): GameStats {
  const totalXp = attempts.reduce((s, a) => s + a.overallScore, 0);
  const level = Math.floor(totalXp / XP_PER_LEVEL) + 1;
  const xpInLevel = totalXp % XP_PER_LEVEL;

  const days = new Set(attempts.map((a) => dayIndex(a.timestamp)));
  const today = dayIndex(Date.now());
  const streak = currentStreak(days, today);
  const best = bestStreak(days);

  const totalWords = attempts.reduce((s, a) => s + a.wordCount, 0);
  const bestScore = attempts.length ? Math.max(...attempts.map((a) => a.overallScore)) : 0;
  const modulesTried = new Set(attempts.map((a) => a.moduleId)).size;

  const badges: Badge[] = [
    { id: "first", emoji: "🌟", title: "Birinchi qadam", hint: "Birinchi mashqni bajar", unlocked: attempts.length > 0 },
    { id: "five", emoji: "🔥", title: "5 ta mashq", hint: "5 ta mashq bajar", unlocked: attempts.length >= 5 },
    { id: "twenty", emoji: "💪", title: "20 ta mashq", hint: "20 ta mashq bajar", unlocked: attempts.length >= 20 },
    { id: "score80", emoji: "🏆", title: "Zo'r natija", hint: "80+ ball to'pla", unlocked: bestScore >= 80 },
    { id: "score95", emoji: "👑", title: "Mukammal", hint: "95+ ball to'pla", unlocked: bestScore >= 95 },
    { id: "streak3", emoji: "📅", title: "3 kun ketma-ket", hint: "3 kun to'xtamay mashq qil", unlocked: best >= 3 },
    { id: "streak7", emoji: "🗓️", title: "Bir hafta", hint: "7 kun to'xtamay mashq qil", unlocked: best >= 7 },
    { id: "explorer", emoji: "🧭", title: "Kashfiyotchi", hint: "Barcha modullarni sinab ko'r", unlocked: moduleCount > 0 && modulesTried >= moduleCount },
    { id: "talker", emoji: "💬", title: "500 so'z", hint: "Jami 500 ta so'z gapir", unlocked: totalWords >= 500 },
  ];

  return {
    totalXp,
    level,
    levelTitle: LEVEL_TITLES[level - 1] ?? LEVEL_TITLES[LEVEL_TITLES.length - 1],
    xpInLevel,
    xpPerLevel: XP_PER_LEVEL,
    streakDays: streak,
    bestStreak: best,
    practicedToday: days.has(today),
    totalAttempts: attempts.length,
    totalWords,
    badges,
    levelProgress: xpInLevel / XP_PER_LEVEL,
    unlockedBadges: badges.filter((b) => b.unlocked).length,
  };
}

File: app/src/main/java/uz/speakingapp/analysis/Gamification.kt
package uz.speakingapp.analysis

import uz.speakingapp.data.db.AttemptEntity
import java.util.TimeZone

/** Bitta nishon (badge). */
data class Badge(
    val id: String,
    val emoji: String,
    val title: String,
    val hint: String,
    val unlocked: Boolean,
)

/** O'yin mexanikalari holati — daraja, XP, kunlik seriya va nishonlar. */
data class GameStats(
    val totalXp: Int,
    val level: Int,
    val levelTitle: String,
    val xpInLevel: Int,
    val xpPerLevel: Int,
    val streakDays: Int,
    val bestStreak: Int,
    val practicedToday: Boolean,
    val totalAttempts: Int,
    val totalWords: Int,
    val badges: List<Badge>,
) {
    val levelProgress: Float
        get() = if (xpPerLevel == 0) 0f else xpInLevel.toFloat() / xpPerLevel

    val unlockedBadges: Int get() = badges.count { it.unlocked }
}

/**
 * Mahalliy natijalardan (Room) motivatsiya ko'rsatkichlarini hisoblaydi.
 * Hech qanday server kerak emas — hammasi offline ishlaydi.
 */
object Gamification {

    private const val XP_PER_LEVEL = 400

    private val LEVEL_TITLES = listOf(
        "Yangi boshlovchi",
        "Mashqchi",
        "Suhbatdosh",
        "Notiq",
        "Usta notiq",
        "Chempion",
    )

    fun compute(attempts: List<AttemptEntity>, moduleCount: Int): GameStats {
        // XP: har bir urinish o'z balliga teng XP beradi.
        val totalXp = attempts.sumOf { it.overallScore }
        val level = totalXp / XP_PER_LEVEL + 1
        val xpInLevel = totalXp % XP_PER_LEVEL

        val days = attempts.map { dayIndex(it.timestamp) }.toSortedSet()
        val today = dayIndex(System.currentTimeMillis())
        val streak = currentStreak(days, today)
        val best = bestStreak(days)

        val totalWords = attempts.sumOf { it.wordCount }
        val bestScore = attempts.maxOfOrNull { it.overallScore } ?: 0
        val modulesTried = attempts.map { it.moduleId }.toSet().size

        val badges = listOf(
            Badge("first", "🌟", "Birinchi qadam", "Birinchi mashqni bajar", attempts.isNotEmpty()),
            Badge("five", "🔥", "5 ta mashq", "5 ta mashq bajar", attempts.size >= 5),
            Badge("twenty", "💪", "20 ta mashq", "20 ta mashq bajar", attempts.size >= 20),
            Badge("score80", "🏆", "Zo'r natija", "80+ ball to'pla", bestScore >= 80),
            Badge("score95", "👑", "Mukammal", "95+ ball to'pla", bestScore >= 95),
            Badge("streak3", "📅", "3 kun ketma-ket", "3 kun to'xtamay mashq qil", best >= 3),
            Badge("streak7", "🗓️", "Bir hafta", "7 kun to'xtamay mashq qil", best >= 7),
            Badge(
                "explorer", "🧭", "Kashfiyotchi", "Barcha modullarni sinab ko'r",
                moduleCount > 0 && modulesTried >= moduleCount,
            ),
            Badge("talker", "💬", "500 so'z", "Jami 500 ta so'z gapir", totalWords >= 500),
        )

        return GameStats(
            totalXp = totalXp,
            level = level,
            levelTitle = LEVEL_TITLES.getOrElse(level - 1) { LEVEL_TITLES.last() },
            xpInLevel = xpInLevel,
            xpPerLevel = XP_PER_LEVEL,
            streakDays = streak,
            bestStreak = best,
            practicedToday = days.contains(today),
            totalAttempts = attempts.size,
            totalWords = totalWords,
            badges = badges,
        )
    }

    /** Mahalliy vaqt mintaqasidagi kun raqami (1970-01-01 dan boshlab). */
    private fun dayIndex(timestamp: Long): Long {
        val offset = TimeZone.getDefault().getOffset(timestamp)
        return Math.floorDiv(timestamp + offset, 86_400_000L)
    }

    /**
     * Joriy seriya: bugundan (yoki kechadan — bugun hali mashq qilmagan bo'lsa)
     * orqaga qarab uzluksiz kunlar soni.
     */
    private fun currentStreak(days: Set<Long>, today: Long): Int {
        var cursor = when {
            days.contains(today) -> today
            days.contains(today - 1) -> today - 1
            else -> return 0
        }
        var count = 0
        while (days.contains(cursor)) {
            count++
            cursor--
        }
        return count
    }

    /** Eng uzun uzluksiz kunlar seriyasi (butun tarix bo'yicha). */
    private fun bestStreak(days: Set<Long>): Int {
        if (days.isEmpty()) return 0
        var best = 1
        var run = 1
        val sorted = days.sorted()
        for (i in 1 until sorted.size) {
            run = if (sorted[i] == sorted[i - 1] + 1) run + 1 else 1
            if (run > best) best = run
        }
        return best
    }
}
